soc/xtensa/intel_adsp/common/CMakeLists.txt

```diff
@@ -100,3 +100,103 @@ add_custom_target(
       ${CMAKE_BINARY_DIR}/zephyr/main.mod
       ${CMAKE_BINARY_DIR}/zephyr/main.mod 2>${NULL_FILE}
 )
+
+
+# Build the rimage tool. We signed the RIMAGE in the building phase now.
+# It will try to find out current rimage tool fisrt. If the rimage tool
+# cannot be found, it will try to fetch the rimage source code and
+# build the rimage tool binary by this script:
+#
+#    soc/xtensa/intel_adsp/tools/build_rimage.sh
+#
+# It will fetch the code of rimage via git submodule update, then
+# rimage tool will be built in modules/audio/sof/rimage/, then it will
+# be copied to ZEPHYR_BASE. While the rimage was built, the following
+# applications/tests will use it directly.
+#
+# The name of the key needed for signed will be specified in the
+# boards/xtensa/{cavs_board}/board.cmake by setting up the variable
+# RIMAGE_CONFIG_SCHEMA. The .toml will be specified directly base on
+# the target board. And these two file will be got also from the
+# modules/audio/sof/rimage/ repo. Ideally, we don't have to do this
+# manually, it will be done automatically durning our building stage.
+#
+# If you want to specify the key and config instead of the default
+# ones, you can also assign them by:
+#   west build -b intel_adsp_cavsxx samples/hello_world
+#     -DRIMAGE_KEY=path_to/xxx.key -DRIMAGE_CONFIG=path_to/xxx.toml
+#
+find_program(RIMAGE rimage
+  PATH ${ZEPHYR_SOF_MODULE_DIR}/rimage/build ${ZEPHYR_BASE})
+
+find_file(RIMAGE_CONFIG_SCHEMA ${RIMAGE_TARGET}.toml
+  PATH ${ZEPHYR_SOF_MODULE_DIR}/rimage/config)
+
+# Try to build the rimage tool and get the toml config.
+if(${RIMAGE} STREQUAL RIMAGE-NOTFOUND
+  OR ${RIMAGE_CONFIG_SCHEMA} STREQUAL RIMAGE_CONFIG_SCHEMA-NOTFOUND)
+
+  find_program(RIMAGE_BUILD_SCRIPT build_rimage.sh
+    PATH ${ZEPHYR_BASE}/soc/xtensa/intel_adsp/tools/)
+    message(WARNING "Start to build rimage tool by ${ZEPHYR_BASE}")
+
+  execute_process(
+    COMMAND
+    ${RIMAGE_BUILD_SCRIPT}
+    ${ZEPHYR_BASE}
+    ${ZEPHYR_SOF_MODULE_DIR}
+    RESULT_VARIABLE return_code
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+    WORKING_DIRECTORY ${ZEPHYR_SOF_MODULE_DIR}/rimage
+  )
+  if(return_code)
+    message(WARNING "build rimage tool failed: ${stderr}")
+  endif()
```

Comment:
@stephanosio @nashif while @tejlmand is on leave, I won't block over this, but this commit looks a bit off to me.

It seems like ExternalProject_add() or something like that should be used instead if you really need this. It's not clear to me why the zephyr build system should be responsible for compiling a host tool, which is what I think rimage is.

Reply:
I agree with @mbolivar-nordic.

If `rimage` (a host tool) is required, downloading or building it should be an independent process from building the Zephyr image.

Please either provide the instruction for download/building the `rimage` in the documentation, or provide a west extension that does this (similar to how Espressif handles their toolchain installation).

In the long term, it may be preferable to add this host tool as part of the Zephyr SDK (see https://github.com/zephyrproject-rtos/sdk-ng/issues/462).

Reply:
Hi @mbolivar-nordic @stephanosio thanks for letting me know the right way to do it. I totally remove the "rimage tool build process", and specify the rimage build documentation link, as @nashif suggested. And I am also looking forward to https://github.com/zephyrproject-rtos/sdk-ng/issues/462. Thanks for the review.